compaction: panic due to index out of bounds

I noticed this while investigating cockroachdb/cockroach#82924.

Node 3 crashed with the following:

```console
panic: runtime error: index out of range [2] with length 2

goroutine 171900 [running]:
panic({0x4cc5e40, 0xc00a25ccd8})
        GOROOT/src/runtime/panic.go:1147 +0x3a8 fp=0xc005ba5590 sp=0xc005ba54d0 pc=0x48a828
runtime.goPanicIndex(0x2, 0x2)
        GOROOT/src/runtime/panic.go:90 +0x75 fp=0xc005ba55d0 sp=0xc005ba5590 pc=0x487bb5
github.com/cockroachdb/pebble/internal/manifest.(*L0Sublevels).InitCompactingFileInfo(0xc00509b170, {0xc003c08b40, 0x1, 0xc00503c510})
        github.com/cockroachdb/pebble/internal/manifest/external/com_github_cockroachdb_pebble/internal/manifest/l0_sublevels.go:721 +0x4be fp=0xc005ba5850 sp=0xc005ba55d0 pc=0x10bf8fe
github.com/cockroachdb/pebble.(*DB).removeInProgressCompaction(0xc000b47200, 0xc00194bb00)
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1200 +0xb7 fp=0xc005ba5be0 sp=0xc005ba5850 pc=0x114b457
github.com/cockroachdb/pebble.(*DB).compact1(0xc000b47200, 0xc00194bb00, 0x0)
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1904 +0x5ef fp=0xc005ba5e80 sp=0xc005ba5be0 pc=0x115028f
github.com/cockroachdb/pebble.(*DB).compact.func1({0x66a5bd8, 0xc005cec2d0})
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1840 +0xad fp=0xc005ba5ee8 sp=0xc005ba5e80 pc=0x114fbad
runtime/pprof.Do({0x66a5b68, 0xc000076088}, {{0xc0000cee40, 0x1074827b0, 0xc001d3e7d0}}, 0xc007482788)
        GOROOT/src/runtime/pprof/runtime.go:40 +0xa3 fp=0xc005ba5f58 sp=0xc005ba5ee8 pc=0x1027ba3
github.com/cockroachdb/pebble.(*DB).compact(0xc0074827d0, 0x11248c6, 0xc001d3e7c0)
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1837 +0x6b fp=0xc005ba5fb8 sp=0xc005ba5f58 pc=0x114facb
github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker·dwrap·13()
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x2e fp=0xc005ba5fe0 sp=0xc005ba5fb8 pc=0x114e38e
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc005ba5fe8 sp=0xc005ba5fe0 pc=0x4bf321
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker
        github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x5c5
```

Node 3 also had a bad block device, so this may have contributed.

Running with Pebble version ad44a62e4d04.

--- a/compaction.go
+++ b/compaction.go
@@ -1650,6 +1650,15 @@ func (d *DB) flush1() (bytesFlushed uint64, err error) {
 			// TODO(peter): untested.
 			d.mu.versions.obsoleteTables = append(d.mu.versions.obsoleteTables, pendingOutputs...)
 			d.mu.versions.incrementObsoleteTablesLocked(pendingOutputs)
+
+			// If logAndApply fails, there is a chance that we ran NewL0Sublevels
+			// successfully, or at least ran it for long enough to mutate
+			// f.{min,max}IntervalIndex on many files. Since that NewL0Sublevels was
+			// on a Version that was not installed, regenerate the current version's
+			// NewL0Sublevels to ensure that all interval indices on FileMetadatas
+			// are correct again. This is necessary to avoid future panics in
+			// InitCompactingFileInfo as well as to aid in correct compaction picking.
+			_ = d.mu.versions.currentVersion().InitL0Sublevels(c.cmp, c.formatKey, d.opts.FlushSplitBytes)
 		}
 	}
 
@@ -2145,6 +2154,10 @@ func (d *DB) compact1(c *compaction, errChannel chan error) (err error) {
 			// TODO(peter): untested.
 			d.mu.versions.obsoleteTables = append(d.mu.versions.obsoleteTables, pendingOutputs...)
 			d.mu.versions.incrementObsoleteTablesLocked(pendingOutputs)
+
+			// See comment in the similar invocation of InitL0Sublevels in flush1()
+			// above, on why this is necessary.
+			_ = d.mu.versions.currentVersion().InitL0Sublevels(c.cmp, c.formatKey, d.opts.FlushSplitBytes)
 		}
 	}
 

--- a/internal/manifest/l0_sublevels.go
+++ b/internal/manifest/l0_sublevels.go
@@ -714,6 +714,16 @@ func (s *L0Sublevels) InitCompactingFileInfo(inProgress []L0Compaction) {
 
 	iter := s.levelMetadata.Iter()
 	for f := iter.First(); f != nil; f = iter.Next() {
+		if invariants.Enabled {
+			if !bytes.Equal(s.orderedIntervals[f.minIntervalIndex].startKey.key, f.Smallest.UserKey) {
+				panic(fmt.Sprintf("f.minIntervalIndex in FileMetadata out of sync with intervals in L0Sublevels: %s != %s",
+					s.formatKey(s.orderedIntervals[f.minIntervalIndex].startKey.key), s.formatKey(f.Smallest.UserKey)))
+			}
+			if !bytes.Equal(s.orderedIntervals[f.maxIntervalIndex+1].startKey.key, f.Largest.UserKey) {
+				panic(fmt.Sprintf("f.maxIntervalIndex in FileMetadata out of sync with intervals in L0Sublevels: %s != %s",
+					s.formatKey(s.orderedIntervals[f.maxIntervalIndex+1].startKey.key), s.formatKey(f.Smallest.UserKey)))
+			}
+		}
 		if !f.Compacting {
 			continue
 		}

